### PR TITLE
Create a generic data item provider for all file based datasources

### PR DESCRIPTION
**Pull request by nyalldawson**
~~(temporarily contains https://github.com/qgis/QGIS/pull/44341)~~

This provider uses the QgsProviderRegistry::querySublayers API to automatically create appropriate browser data items for all
file based sources, regardless of the underlying provider (be it mdal, gdal, ogr, pdal or ept).

This allows us to merge sources which can be handled by multiple different providers into single container items in the browser,
so e.g. instead of seeing a .nc file appear twice in the browser (once with a mesh icon and once with a raster icon), we now
only see the file ONCE and expanding it out shows BOTH the mesh and raster sublayers. Similarly with all other mixed type formats, such as GeoPDF files (which may contain a mix of raster and vector layers), KML/KMZ,... etc.

And since we've one generic provider now, we can remove the mdal, gdal, ogr, pdal and ept data item providers entirely, as we now get their functionality "for free"!

### CODE REVIEW

File: src/core/browser/qgsfilebaseddataitemprovider.cpp [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+
+  // test that file is valid with OGR
+  if ( OGRGetDriverCount() == 0 )
+  {
+    OGRRegisterAll();
+  }
+  // do not print errors, but write to debug
+  CPLPushErrorHandler( CPLQuietErrorHandler );
+  CPLErrorReset();
+  gdal::dataset_unique_ptr hDS( GDALOpenEx( path().toUtf8().constData(), GDAL_OF_VECTOR, nullptr, nullptr, nullptr ) );
```
Comment (elpaso) [inline src/core/browser/qgsfilebaseddataitemprovider.cpp:190]: Can you add the READ ONLY flag to openex?

File: src/core/browser/qgsfilebaseddataitemprovider.cpp [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+}
+
+QgsDataItem *QgsFileBasedDataItemProvider::createDataItem( const QString &pathIn, QgsDataItem *parentItem )
+{
+  QString path( pathIn );
+  if ( path.isEmpty() )
+    return nullptr;
+
+  const QFileInfo info( path );
+  QString suffix = info.suffix().toLower();
```
Comment (elpaso) [inline src/core/browser/qgsfilebaseddataitemprovider.cpp:236]:
```suggestion
  const QString suffix = info.suffix().toLower();
```
?

File: src/core/browser/qgsfilebaseddataitemprovider.cpp [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+        QString encodedUri = uq.toString();
+        return new QgsVectorTileLayerItem( parentItem, name, path, encodedUri );
+      }
+      else
+      {
+        // handled by WMS provider
+        QUrlQuery uq;
+        uq.addQueryItem( QStringLiteral( "type" ), QStringLiteral( "mbtiles" ) );
+        uq.addQueryItem( QStringLiteral( "url" ), QUrl::fromLocalFile( path ).toString() );
+        QString encodedUri = uq.toString();
```
Comment (elpaso) [inline src/core/browser/qgsfilebaseddataitemprovider.cpp:296]:
```suggestion
        const QString encodedUri = uq.toString();
```

File: src/core/browser/qgsfilebaseddataitemprovider.cpp [hunk 4]
```diff
@@ -0,0 +1,353 @@
...
+    QgsMbTiles reader( path );
+    if ( reader.open() )
+    {
+      if ( reader.metadataValue( QStringLiteral( "format" ) ) == QLatin1String( "pbf" ) )
+      {
+        // these are vector tiles
+        QUrlQuery uq;
+        uq.addQueryItem( QStringLiteral( "type" ), QStringLiteral( "mbtiles" ) );
+        uq.addQueryItem( QStringLiteral( "url" ), path );
+        QString encodedUri = uq.toString();
```
Comment (elpaso) [inline src/core/browser/qgsfilebaseddataitemprovider.cpp:287]:
```suggestion
        const QString encodedUri = uq.toString();
```

File: src/core/browser/qgsfilebaseddataitemprovider.h [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+
+    /**
+     * Constructor for QgsProviderSublayerItem.
+     * \param parent parent item
+     * \param name data item name (this should match either the layer's name or the filename of a single-layer file)
+     * \param details sublayer details
+     * \param isFile set to TRUE if the item represents a single-layer file, or FALSE if the item is a sublayer from a
+     * collection item.
+     */
+    QgsProviderSublayerItem( QgsDataItem *parent, const QString &name, const QgsProviderSublayerDetails &details, bool isFile );
```
Comment (elpaso) [inline src/core/browser/qgsfilebaseddataitemprovider.h:57]: I'm not really sure but I think I'd prefer an `isSubLayer` flag instead of `isFile`.

**Review by elpaso — APPROVED** [review]
This is a great step forward, thanks!

I was just wondering how much would it take to remove  gpkg and SL "special" handling, but I'm not sure it worth the effort.

File: src/core/browser/qgsfilebaseddataitemprovider.cpp (hunk 2 above)
Comment (nyalldawson) [reply src/core/browser/qgsfilebaseddataitemprovider.cpp:236]: it's modified in some code paths, so isn't const

File: src/core/browser/qgsfilebaseddataitemprovider.h (hunk 5 above)
Comment (nyalldawson) [reply src/core/browser/qgsfilebaseddataitemprovider.h:57]: Can i leave as is for the moment? I'm wanting to add some type of flag to QgsDataItem later to reflect that the item is a file, and use this in the gui providers to ensure that the file related operations (such as "file properties") are always shown for these items, regardless of their subclass or provider.

**Comment by nyalldawson** [comment]
@elpaso
> I was just wondering how much would it take to remove gpkg and SL "special" handling, but I'm not sure it worth the effort.

Well, I **think** we could safely remove the geopackage data item classes and let the generic item handle these without regressions, and only leave the geopackage gui provider for the special handling. But given the heavy investment in gpkg data items I don't want to do that until I'm confident there's no regressions.

Spatialite's a bit trickier though.... I'd rather not go down that rabbit hole :laughing:

**Pull request merged**